hoomd/mpcd/CollisionMethod.cc

```diff
@@ -48,15 +85,964 @@ void mpcd::CollisionMethod::collide(uint64_t timestep)
+            // loop over all molecules
+            unsigned int n_particles_local = m_pdata->getN() + m_pdata->getNGhosts();
+            for (unsigned int ibody = 0;
+                 ibody < nmol && !(invalid_center_of_mass || rigid_types.empty());
+                 ibody++)
+                {
+                // get central particle tag from first particle in molecule
+                assert(h_molecule_len.data[ibody] > 0);
+                unsigned int first_idx = h_molecule_list.data[molecule_indexer(0, ibody)];
+                if (first_idx > n_particles_local)
+                    continue;
+                unsigned int central_tag = h_body.data[first_idx];
+                unsigned int central_idx = h_rtag.data[central_tag];
+                if (central_idx >= n_particles_local)
+                    continue;
+
+                // only do molecules whose type hasn't been checked before
+                unsigned int type = __scalar_as_int(h_postype.data[central_idx].w);
+                if (rigid_types.find(type) == rigid_types.end())
+                    {
+                    continue;
+                    }
+                else
+                    {
+                    rigid_types.erase(rigid_types.find(type));
+                    }
+
+                // find center of mass position of particle
+                Scalar3 center_of_mass = make_scalar3(0.0, 0.0, 0.0);
+                Scalar mass_sum = Scalar(0.0);
+                Scalar center_mass = h_velocity.data[central_idx].w;
+                for (unsigned int constituent_index = 0;
+                     constituent_index < h_molecule_len.data[ibody];
+                     ++constituent_index)
+                    {
+                    unsigned int idxj
+                        = h_molecule_list.data[molecule_indexer(constituent_index, ibody)];
+                    assert(idxj < m_pdata->getN() + m_pdata->getNGhosts());
+                    if (idxj == central_idx)
+                        continue;
+
+                    // get relative position and mass of constituent
+                    unsigned int idx_in_body = h_molecule_order.data[idxj] - 1;
+                    Scalar3 local_pos(h_body_pos.data[h_body_idx(type, idx_in_body)]);
+                    Scalar mass = h_velocity.data[idxj].w;
+                    // add to accumulating center of mass
+                    center_of_mass.x += local_pos.x * mass;
+                    center_of_mass.y += local_pos.y * mass;
+                    center_of_mass.z += local_pos.z * mass;
+                    mass_sum += mass;
+                    }
+                // check if center of mass in body frame is (0, 0, 0)
+                bool in_tol = (center_of_mass.x >= -1.0 * tol && center_of_mass.x <= tol)
+                              && (center_of_mass.y >= -1.0 * tol && center_of_mass.y <= tol)
+                              && (center_of_mass.z >= -1.0 * tol && center_of_mass.z <= tol);
```

Comment:
Please use a `-` rather than `-1.0 *` throughout.
```suggestion
                bool in_tol = (center_of_mass.x >= -tol && center_of_mass.x <= tol)
                              && (center_of_mass.y >= -tol && center_of_mass.y <= tol)
                              && (center_of_mass.z >= -tol && center_of_mass.z <= tol);
```
Alternatively, use `fabs`